pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/conditions/Unsuccessful.groovy

```diff
@@ -46,16 +47,14 @@ class Unsuccessful extends BuildCondition {
 
     @Override
     boolean meetsCondition(@Nonnull WorkflowRun r, Object context, Throwable error) {
-        return isNotNull(r) || ! is(r, Result.SUCCESS)
```

Comment:
#300 had a bug; this should have been `isNotNull(r) && ! is(r, Result.SUCCESS)`. The previous code made the condition run even if the build is successful. This was already reported as [JENKINS-55476](https://issues.jenkins-ci.org/browse/JENKINS-55476), and I've tweaked the tests so they'll catch any regressions here in the future.